### PR TITLE
nRF52/aes128_ctr

### PR DESCRIPTION
**Pull request by niklasad1**
Hi,
This is copy-paste from nRF51 perhaps I should later check exactly difference between nRF51 and nRF52 and have some common implementations or similar.

### CODE REVIEW

**Comment by ppannuto** [comment]
Would it make more sense to make a Nordic crate, or an NRF5X crate, and
simply have the 51 and 52 re export common things? I imagine the
duplication is going to just keep stacking up :/

**Comment by niklasad1** [comment]
I think a NRF5X crate would be good to start with.
We already got a lot of code duplication so it's about time IMO.

**Comment by alevy** [comment]
Should we merge this in the meantime? Or close it and do an NRF5x crate instead already?

**Comment by niklasad1** [comment]
I think we should merge this now because I'm leaving for Cyprus early tomorrow and I don't have time to work on NRF5x crate for ~1 week.

File: chips/nrf52/src/aes.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
+//! AES128-CTR Driver
+//!
+//! Provides a simple driver for userspace applications to encrypt and decrypt messages
+//! using aes128-ctr mode on top of aes128-ecb
```
Comment (bradjc) [inline chips/nrf52/src/aes.rs]: Add a ".".

File: chips/nrf52/src/aes.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+//!
+//! The initial counter configured according to the counter received from the user application.
+//! The capsule is invoked as follows:
+//!     - the key has been configured
```
Comment (bradjc) [inline chips/nrf52/src/aes.rs]: Need an empty comment line before this list for the markdown to render correctly.

File: chips/nrf52/src/aes.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+//!     - the entire buffer has been encrypted
+//!     - the entire buffer has been decrypted
+//!
+//! The buffer is also sliced in chips at the moment and some un-necessary
+//! static mut...
+//!
+//! FIXME:
+//!     - maybe move some stuff to capsule instead
```
Comment (bradjc) [inline chips/nrf52/src/aes.rs]: Need an empty comment line before this list here too.

File: chips/nrf52/src/aes.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+//!     - INIT_CTR can be replaced with TakeCell
+//!     - ECB_DATA must be a static mut [u8]
+//!       and can't be located in the struct
+//!     - PAYLOAD size is restricted to 128 bytes
+//!
+//! Author: Niklas Adolfsson <niklasadolfsson1@gmail.com>
+//! Author: Fredrik Nilsson <frednils@student.chalmers.se>
+//! Date: April 21, 2017
```
Comment (bradjc) [inline chips/nrf52/src/aes.rs]: Make these 3 lines a list. Otherwise they render as an odd paragraph.

File: chips/nrf52/src/aes.rs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+// data to replace TakeCell initial counter in the capsule
+static mut INIT_CTR: [u8; 16] = [0; 16];
+
+pub struct AesECB {
+    regs: *const peripheral_registers::AESECB_REGS,
+    client: Cell<Option<&'static kernel::hil::symmetric_encryption::Client>>,
+    ctr: Cell<[u8; 16]>,
+    // input either plaintext or ciphertext to be encrypted or decrypted
+    input: TakeCell<'static, [u8]>,
+    // keystream to be XOR:ed with the input
```
Comment (bradjc) [inline chips/nrf52/src/aes.rs]: Triple "///" these comments so they show up in the docs.

**Comment by niklasad1** [comment]
I will close this PR because  [https://github.com/helena-project/tock/pull/585](https://github.com/helena-project/tock/pull/585) solves this in a better way!

**Pull request closed without merging**